pallets/baby-liminal/src/lib.rs

```diff
@@ -222,11 +268,18 @@ pub mod pallet {
                 Error::<T>::IdentifierAlreadyInUse
             );
 
+            let deposit = T::VerificationKeyDepositAmount::get();
+            T::Currency::reserve(&who, deposit).map_err(|_| Error::<T>::CannotAffordDeposit)?;
+
             VerificationKeys::<T>::insert(
                 identifier,
                 BoundedVec::try_from(key).unwrap(), // must succeed since we've just check length
             );
 
+            // will never overwrite anything since we already check the VerificationKeys map
+            VerificationKeyOwners::<T>::insert(identifier, &who);
```

Comment:
Why do we require that `identifier`s are unique across whole chain? Can't we enrich it with `origin` (or maybe even better 
- an account_id of contract that called the method) and "namespace" it? Why is it that people can "claim" specific identifiers for themselves and prevent others from using it?

Reply:
There is  2^32 unique identifier possibilities, one per every pair of humans on Earth. I just don't see a use-case in increasing it even further.

Reply:
That's a fair point. I was thinking about other people that would want to use similar technic to derive their keys: `[b'T', b'E', b'S', b'T']` or something.

But this raises another point -- how can people find which keys are taken? Should they use some source of random data for this? If they try to use a mnemonic they could try using something that is already registered. Maybe pallet should be the one to tell under what identifier their key was registered - i.e. turn the dependencies around.

Reply:
I have slight preference to @deuszx 's vision, especially that some people might copy Shielder and want to use the same key identifiers

Reply:
if we make a  (AccountId, [u;4]) tuple an identifier, that this same identifier would need to be passed from the contracts when verifying.  Let's be honest here, we will be the only ones ever registering those keys.

Reply:
I'm fine with extending 4 bytes to more (8/16)

Reply:
:ok_hand:

Reply:
It would be a non-issue if we made it a tupple with _caller's_ `AccountId` - every contract would have its own and we would be passing that automatically -- or rather, read the caller's address in the chain extension. That's IMO the best solution for this.

Reply:
I don't understand this comment - what would be a non-issue?

Reply:
I still don't understand - are you proposing to always be automatically passing the contract callers `AccountId` as an argument to the pallet's functions that need a key identifier? That makes little sense to me, there isn't a 1:1 mapping there, key is registered once and used by many. It's as if every user of the Tornado had to register own key - and for every relation.

Reply:
No, it would be used later on by a contract (in our case - `shielder`) so it's registered and used from the same context every time. If someone else wants to deploy their own shielder they can do that and keys won't conflict as the same key will be registered and used from a different context - different contract. Users aren't using the key directly, it's a contract that does it when they post a txn.

Reply:
I am completely lost here. Keys are unique for relations, why are we discussing confounding them with accountIds ? E.g. why do I need to keep track of some account id to deposit or withdraw funds?

Reply:
This is what I commented on at the very beginning of this thread. Look at how we're choosing the key identifiers: `[b'D', b'E', b'P', b'O', b'S', b'I', b'T']` so not only we're making it "first is the winner" when it comes to taking the keys but also we're making it impossible for users to "just clone" our shielder setup b/c they will have to change all the identifiers under which they're registering the keys.

Reply:
no they don't, they can the same keys that were already registered. I'd argue this is simpler, easier less error-prone and generally design limiting state space size and duplication in blockchain setting is a good thing